optgen: use strictly typed variable references when possible

mgartner opened the pull request:
#### optgen: use strictly typed variable references when possible

This commit changes the code generated by optgen such that strictly
typed variables are passed to custom functions when possible. For
example, consider the following match pattern and custom function:

    (Select $select:(Select $input:* $filters:*)) &
      (ValidSelect $select)

    func (c *CustomFuncs) ValidSelect (s *memo.SelectExpr) { }

Previously, the generated Go code would not compile because the
`$select` variable was typed as `memo.RelExpr`. Now, the generated code
will replace reference to `$select` with a synthesized variable typed as
`*memo.SelectExpr`. Note that this is only possible when there is a
single node expression in the match pattern, e.g., in
`$input:(SemiJoin | AntiJoin)` the `$input` variable cannot be strictly
typed.

Release note: None

#### opt: use strict types with custom functions

Optgen now generates code that passes strictly types variables to
custom functions when the variables are bound to a single type of
expression. This commit capitalizes on this feature and updates the
types of custom function arguments to be stricter than the
`memo.RelExpr` interface when possible.

Release note: None

mgartner committed:
optgen: use strictly typed variable references when possible

This commit changes the code generated by optgen such that strictly
typed variables are passed to custom functions when possible. For
example, consider the following match pattern and custom function:

    (Select $select:(Select $input:* $filters:*)) &
      (ValidSelect $select)

    func (c *CustomFuncs) ValidSelect (s *memo.SelectExpr) { }

Previously, the generated Go code would not compile because the
`$select` variable was typed as `memo.RelExpr`. Now, the generated code
will replace reference to `$select` with a synthesized variable typed as
`*memo.SelectExpr`. Note that this is only possible when there is a
single node expression in the match pattern, e.g., in
`$input:(SemiJoin | AntiJoin)` the `$input` variable cannot be strictly
typed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54293)
<!-- Reviewable:end -->

mgartner commented:
#### Note

I did this mostly as an exercise to better understand optgen. I'm not convinced it's something that we need to do. It might be worthwhile if we've had to come up with workarounds when writing rules in the past due to this limitation. I've only run into this when I was trying to write a rule an in unidiomatic way.

And, if we do want to support this, I'm not sure if my implementation is the best.

mgartner commented on pkg/sql/opt/optgen/cmd/optgen/testdata/factory:
```diff
@@ -82,34 +82,37 @@ func (_f *Factory) ConstructSelect(
+								_f.DynamicConstruct(
+									input.Op(),
+									_f.ConstructSelect(
+										_select,
```
Notice that the first argument to `_f.ConstructSelect` is now `_select`, rather than `left`. This highlights the change in generated code.

RaduBerinde reviewed (commented):
There are a lot of test failures, something's not working right.

I like this, I don't see a downside. You can still define the custom function using `RelExpr` (and all existing funcs still work), but if you have something specific to an operator you can use that specific type and save a cast. Would be good to take a quick look to see if there are any existing functions that would be reasonable to change like that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54293#-:-MH7jHBUCUgQJEWoDBXw:b-db983e)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

mgartner committed:
opt: use strict types with custom functions

Optgen now generates code that passes strictly types variables to
custom functions when the variables are bound to a single type of
expression. This commit capitalizes on this feature and updates the
types of custom function arguments to be stricter than the
`memo.RelExpr` interface when possible.

Release note: None

mgartner commented:
This is ready for another look. I fixed the bug (it was aliasing with the type-asserted variable in not-match expressions, e.g., `$input:^(Scan)`), cleaned up the tests, and added another commit that removes type assertions that are no longer necessary due to this change (I _think_ I found all of them).

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  cool!

> Reviewed 1 of 4 files at r1, 3 of 3 files at r2, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54293#-:-MHqB62k733xG53MNDEk:b-4v09)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54293#-:-MHrFNgJ4lggoBBbQHU6:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+
